File: permutation2.py
class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
    def copy(self, origin):
        result = []
        for i in origin:
            result.append(i)
        return result

    def get_next(self, prev):
        suffix_head = len(prev) - 1
        cur = self.copy(prev)
        while (suffix_head >= 0 and cur[suffix_head - 1] >= cur[suffix_head]):
            suffix_head -= 1
        if (suffix_head <= 0):
            return None
        i = len(cur) - 1
        while (cur[i] <= cur[suffix_head - 1]):
            i -= 1
        temp = cur[suffix_head - 1]
        cur[suffix_head - 1] = cur[i]
        cur[i] = temp
        for i in range(suffix_head, len(cur) - 1):
            for j in range(i + 1, len(cur)):
                if cur[i] > cur[j]:
                    temp = cur[i]
                    cur[i] = cur[j]
                    cur[j] = temp
        return cur


    def permute(self, num):
        result = []
        num.sort()
        result.append(num)
        nt = self.get_next(num)
        while (nt):
            result.append(nt)
            nt = self.get_next(nt)
        return result

File: permutation2.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
  public:
    vector<int> getNext(vector<int> prev, bool &finished) {
      vector<int> cur(prev);
      int suffixHead = cur.size() - 1;
      while (suffixHead >= 0 && cur[suffixHead - 1] >= cur[suffixHead])
        suffixHead--;
      // cout << "suffixHead: " << suffixHead + 1 << endl;
      if (suffixHead == 0) {
        finished = true;
        return cur;
      }
      int pivot = suffixHead - 1;
      int candidateInd = cur.size() - 1;
      while (cur[candidateInd] <= cur[pivot])
        candidateInd--;
      // cout << "candidateInd" << candidateInd + 1 << endl;
      cur[candidateInd] ^= cur[pivot];
      cur[pivot] ^= cur[candidateInd];
      cur[candidateInd] ^= cur[pivot];
      for (int i = suffixHead; i < (cur.size() + suffixHead) / 2; i++) {
        int j = cur.size() - 1 + suffixHead - i;
        // cout << "swap: " << i + 1 << " and " << j + 1 << endl;
        cur[i] ^= cur[j];
        cur[j] ^= cur[i];
        cur[i] ^= cur[j];
      }
      finished = false;
      return cur;
    }
    vector<vector<int> > permute(vector<int> &num) {
      vector<vector<int> > result;
      vector<int> n(num);
      for (int i = 0; i < n.size() - 1; i++)
        for (int j = i + 1; j < n.size(); j++) {
          if (n[i] > n[j]) {
            n[i] ^= n[j];
            n[j] ^= n[i];
            n[i] ^= n[j];
          }
        }
      result.push_back(n);
      bool finished = false;
      vector<int> cur;
      while (!finished) {
        cur = getNext(result.back(), finished);
        if (!finished)
          result.push_back(cur);
      }
      return result;
    }
};

int main() {
  Solution sol;
  vector<int> num {1};
  vector<vector<int> > result = sol.permute(num);
  for (int i = 0; i < result.size(); i++) {
    cout << "[ ";
    for (int j = 0; j < result[i].size(); j++) 
      cout << result[i][j];
    cout << " ]" << endl;
  }
}
